server/src/main/java/org/elasticsearch/common/bytes/ReleasableBytesReference.java

```diff
@@ -34,41 +33,44 @@
  * An extension to {@link BytesReference} that requires releasing its content. This
  * class exists to make it explicit when a bytes reference needs to be released, and when not.
  */
-public final class ReleasableBytesReference extends AbstractRefCounted implements Releasable, BytesReference {
+public final class ReleasableBytesReference implements Releasable, BytesReference {
 
     public static final Releasable NO_OP = () -> {};
     private final BytesReference delegate;
-    private final Releasable releasable;
+    private final AbstractRefCounted refCounted;
 
     public ReleasableBytesReference(BytesReference delegate, Releasable releasable) {
-        super("bytes-reference");
         this.delegate = delegate;
-        this.releasable = releasable;
+        this.refCounted = new RefCountedReleasable(releasable);
+    }
+
+    private ReleasableBytesReference(BytesReference delegate, AbstractRefCounted refCounted) {
+        this.delegate = delegate;
+        this.refCounted = refCounted;
     }
 
     public static ReleasableBytesReference wrap(BytesReference reference) {
         return new ReleasableBytesReference(reference, NO_OP);
     }
 
-    @Override
-    protected void closeInternal() {
-        Releasables.close(releasable);
+    public int refCount() {
+        return refCounted.refCount();
     }
 
     public ReleasableBytesReference retain() {
-        incRef();
+        refCounted.incRef();
         return this;
     }
 
     public ReleasableBytesReference retainedSlice(int from, int length) {
         BytesReference slice = delegate.slice(from, length);
-        incRef();
-        return new ReleasableBytesReference(slice, this);
+        refCounted.incRef();
```

Comment:
I would prefer to move this call to the private `ReleasableBytesReference` constructor.